### PR TITLE
[SPARK-41976][SQL] Improve error message for `INDEX_NOT_FOUND`

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to improve error message for `INDEX_NOT_FOUND`.

### Why are the changes needed?

Make the error message more clear and proper.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Fix UT and `./build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite*`

### CODE REVIEW

**Comment by itholic** [comment]
Test passed.
cc @MaxGekk @srielau FYI

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala [hunk 1]
```diff
@@ -224,7 +224,11 @@ private object PostgresDialect extends JdbcDialect with SQLConfHelper {
         sqlException.getSQLState match {
           // https://www.postgresql.org/docs/14/errcodes-appendix.html
           case "42P07" => throw new IndexAlreadyExistsException(message, cause = Some(e))
-          case "42704" => throw new NoSuchIndexException(message, cause = Some(e))
+          case "42704" =>
+            // The message is: Failed to drop index indexName in tableName
+            val regex = "(?s)Failed to drop index (.*) in".r
```
Comment (srielau) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:229]: Seems we may be able to pick up the qualifiers as well ?

**Review by srielau — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala (hunk 1 above)
Comment (itholic) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:229]: Do you mean `tableName` so maybe do we want to show the error message like:

`Cannot find the index <indexName> on table <tableName>.`

?

Comment (srielau) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:229]: Yeah..

Comment (itholic) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala]: Thanks! Just update the message

**Comment by itholic** [comment]
Also cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**